File: FirebaseDatabase/Sources/Snapshot/FSnapshotUtilities.m
/*
 * Copyright 2017 Google
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "FSnapshotUtilities.h"
#import "FChildrenNode.h"
#import "FCompoundWrite.h"
#import "FConstants.h"
#import "FEmptyNode.h"
#import "FLLRBValueNode.h"
#import "FLeafNode.h"
#import "FMaxNode.h"
#import "FNamedNode.h"
#import "FUtilities.h"
#import "FValidation.h"

@implementation FSnapshotUtilities

+ (id<FNode>)nodeFrom:(id)val {
    return [FSnapshotUtilities nodeFrom:val priority:nil];
}

+ (id<FNode>)nodeFrom:(id)val priority:(id)priority {
    return [FSnapshotUtilities nodeFrom:val
                               priority:priority
                     withValidationFrom:@"nodeFrom:priority:"];
}

+ (id<FNode>)nodeFrom:(id)val withValidationFrom:(NSString *)fn {
    return [FSnapshotUtilities nodeFrom:val priority:nil withValidationFrom:fn];
}

+ (id<FNode>)nodeFrom:(id)val
              priority:(id)priority
    withValidationFrom:(NSString *)fn {
    return [FSnapshotUtilities nodeFrom:val
                               priority:priority
                     withValidationFrom:fn
                                atDepth:0
                                   path:[[NSMutableArray alloc] init]];
}

+ (id<FNode>)nodeFrom:(id)val
              priority:(id)aPriority
    withValidationFrom:(NSString *)fn
               atDepth:(int)depth
                  path:(NSMutableArray *)path {
    @autoreleasepool {
        return [FSnapshotUtilities internalNodeFrom:val
                                           priority:aPriority
                                 withValidationFrom:fn
                                            atDepth:depth
                                               path:path];
    }
}

+ (id<FNode>)internalNodeFrom:(id)val
                     priority:(id)aPriority
           withValidationFrom:(NSString *)fn
                      atDepth:(int)depth
                         path:(NSMutableArray *)path {

    if (depth > kFirebaseMaxObjectDepth) {
        NSRange range;
        range.location = 0;
        range.length = 100;
        NSString *pathString =
            [[path subarrayWithRange:range] componentsJoinedByString:@"."];
        @throw [[NSException alloc]
            initWithName:@"InvalidFirebaseData"
                  reason:[NSString stringWithFormat:
                                       @"(%@) Max object depth exceeded: %@...",
                                       fn, pathString]
                userInfo:nil];
    }

    if (val == nil || val == [NSNull null]) {
        // Null is a valid type to store
        return [FEmptyNode emptyNode];
    }

    [FValidation validateFrom:fn isValidPriorityValue:aPriority withPath:path];
    id<FNode> priority = [FSnapshotUtilities nodeFrom:aPriority];

    id value = val;
    BOOL isLeafNode = NO;

    if ([value isKindOfClass:[NSDictionary class]]) {
        NSDictionary *dict = val;
        if (dict[kPayloadPriority] != nil) {
            id rawPriority = [dict objectForKey:kPayloadPriority];
            [FValidation validateFrom:fn
                 isValidPriorityValue:rawPriority
                             withPath:path];
            priority = [FSnapshotUtilities nodeFrom:rawPriority];
        }

        if (dict[kPayloadValue] != nil) {
            value = [dict objectForKey:kPayloadValue];
            if ([FValidation validateFrom:fn
                         isValidLeafValue:value
                                 withPath:path]) {
                isLeafNode = YES;
            } else {
                @throw [[NSException alloc]
                    initWithName:@"InvalidLeafValueType"
                          reason:[NSString stringWithFormat:
                                               @"(%@) Invalid data type used "
                                               @"with .value. Can only use "
                                                "NSString and NSNumber or be "
                                                "null. Found %@ instead.",
                                               fn, [[value class] description]]
                        userInfo:nil];
            }
        }
    }

    if ([FValidation validateFrom:fn isValidLeafValue:value withPath:path]) {
        isLeafNode = YES;
    }

    if (isLeafNode) {
        return [[FLeafNode alloc] initWithValue:value withPriority:priority];
    }

    // Unlike with JS, we have to handle the dictionary and array cases
    // separately.
    if ([value isKindOfClass:[NSDictionary class]]) {
        NSDictionary *dval = (NSDictionary *)value;
        NSMutableDictionary *children =
            [NSMutableDictionary dictionaryWithCapacity:dval.count];

        // Avoid creating a million newPaths by appending to old one
        for (id keyId in dval) {
            [FValidation validateFrom:fn
                   validDictionaryKey:keyId
                             withPath:path];
            NSString *key = (NSString *)keyId;

            if (![key hasPrefix:kPayloadMetadataPrefix]) {
                [path addObject:key];
                id<FNode> childNode = [FSnapshotUtilities nodeFrom:dval[key]
                                                          priority:nil
                                                withValidationFrom:fn
                                                           atDepth:depth + 1
                                                              path:path];
                [path removeLastObject];

                if (![childNode isEmpty]) {
                    children[key] = childNode;
                }
            }
        }

        if ([children count] == 0) {
            return [FEmptyNode emptyNode];
        } else {
            FImmutableSortedDictionary *childrenDict =
                [FImmutableSortedDictionary
                    fromDictionary:children
                    withComparator:[FUtilities keyComparator]];
            return [[FChildrenNode alloc] initWithPriority:priority
                                                  children:childrenDict];
        }
    } else if ([value isKindOfClass:[NSArray class]]) {
        NSArray *aval = (NSArray *)value;
        NSMutableDictionary *children =
            [NSMutableDictionary dictionaryWithCapacity:aval.count];

        for (int i = 0; i < [aval count]; i++) {
            NSString *key = [NSString stringWithFormat:@"%i", i];
            [path addObject:key];
            id<FNode> childNode =
                [FSnapshotUtilities nodeFrom:[aval objectAtIndex:i]
                                    priority:nil
                          withValidationFrom:fn
                                     atDepth:depth + 1
                                        path:path];
            [path removeLastObject];

            if (![childNode isEmpty]) {
                children[key] = childNode;
            }
        }

        if ([children count] == 0) {
            return [FEmptyNode emptyNode];
        } else {
            FImmutableSortedDictionary *childrenDict =
                [FImmutableSortedDictionary
                    fromDictionary:children
                    withComparator:[FUtilities keyComparator]];
            return [[FChildrenNode alloc] initWithPriority:priority
                                                  children:childrenDict];
        }
    } else {
        NSRange range;
        range.location = 0;
        range.length = MIN(path.count, 50);
        NSString *pathString =
            [[path subarrayWithRange:range] componentsJoinedByString:@"."];

        @throw [[NSException alloc]
            initWithName:@"InvalidFirebaseData"
                  reason:[NSString
                             stringWithFormat:
                                 @"(%@) Cannot store object of type %@ at %@. "
                                  "Can only store objects of type NSNumber, "
                                  "NSString, NSDictionary, and NSArray.",
                                 fn, [[value class] description], pathString]
                userInfo:nil];
    }
}

+ (FCompoundWrite *)compoundWriteFromDictionary:(NSDictionary *)values
                             withValidationFrom:(NSString *)fn {
    FCompoundWrite *compoundWrite = [FCompoundWrite emptyWrite];

    NSMutableArray *updatePaths =
        [NSMutableArray arrayWithCapacity:values.count];
    for (NSString *keyId in values) {
        id value = values[keyId];
        [FValidation validateFrom:fn
            validUpdateDictionaryKey:keyId
                           withValue:value];

        FPath *path = [FPath pathWithString:keyId];
        id<FNode> node = [FSnapshotUtilities nodeFrom:value
                                   withValidationFrom:fn];

        [updatePaths addObject:path];
        compoundWrite = [compoundWrite addWrite:node atPath:path];
    }

    // Check that the update paths are not descendants of each other.
    [updatePaths
        sortUsingComparator:^NSComparisonResult(FPath *left, FPath *right) {
          return [left compare:right];
        }];
    FPath *prevPath = nil;
    for (FPath *path in updatePaths) {
        if (prevPath != nil && [prevPath contains:path]) {
            @throw [[NSException alloc]
                initWithName:@"InvalidFirebaseData"
                      reason:[NSString stringWithFormat:
                                           @"(%@) Invalid path in object. Path "
                                           @"(%@) is an ancestor of (%@).",
                                           fn, prevPath, path]
                    userInfo:nil];
        }
        prevPath = path;
    }

    return compoundWrite;
}

+ (void)validatePriorityNode:(id<FNode>)priorityNode {
    assert(priorityNode != nil);
    if (priorityNode.isLeafNode) {
        id val = priorityNode.val;
        if ([val isKindOfClass:[NSDictionary class]]) {
            NSDictionary *valDict __unused = (NSDictionary *)val;
            NSAssert(valDict[kServerValueSubKey] != nil,
                     @"Priority can't be object unless it's a deferred value");
        } else {
            NSString *jsType __unused = [FUtilities getJavascriptType:val];
            NSAssert(jsType == kJavaScriptString || jsType == kJavaScriptNumber,
                     @"Priority of unexpected type.");
        }
    } else {
        NSAssert(priorityNode == [FMaxNode maxNode] || priorityNode.isEmpty,
                 @"Priority of unexpected type.");
    }
    // Don't call getPriority() on MAX_NODE to avoid hitting assertion.
    NSAssert(priorityNode == [FMaxNode maxNode] ||
                 priorityNode.getPriority.isEmpty,
             @"Priority nodes can't have a priority of their own.");
}

+ (void)appendHashRepresentationForLeafNode:(FLeafNode *)leafNode
                                   toString:(NSMutableString *)string
                                hashVersion:(FDataHashVersion)hashVersion {
    NSAssert(hashVersion == FDataHashVersionV1 ||
                 hashVersion == FDataHashVersionV2,
             @"Unknown hash version: %lu", (unsigned long)hashVersion);
    if (!leafNode.getPriority.isEmpty) {
        [string appendString:@"priority:"];
        [FSnapshotUtilities
            appendHashRepresentationForLeafNode:leafNode.getPriority
                                       toString:string
                                    hashVersion:hashVersion];
        [string appendString:@":"];
    }

    NSString *jsType = [FUtilities getJavascriptType:leafNode.val];
    [string appendString:jsType];
    [string appendString:@":"];

    if (jsType == kJavaScriptBoolean) {
        NSString *boolString =
            [leafNode.val boolValue] ? kJavaScriptTrue : kJavaScriptFalse;
        [string appendString:boolString];
    } else if (jsType == kJavaScriptNumber) {
        NSString *numberString =
            [FUtilities ieee754StringForNumber:leafNode.val];
        [string appendString:numberString];
    } else if (jsType == kJavaScriptString) {
        if (hashVersion == FDataHashVersionV1) {
            [string appendString:leafNode.val];
        } else {
            NSAssert(hashVersion == FDataHashVersionV2,
                     @"Invalid hash version found");
            [FSnapshotUtilities appendHashV2RepresentationForString:leafNode.val
                                                           toString:string];
        }
    } else {
        [NSException raise:NSInvalidArgumentException
                    format:@"Unknown value for hashing: %@", leafNode];
    }
}

+ (void)appendHashV2RepresentationForString:(NSString *)string
                                   toString:(NSMutableString *)mutableString {
    string = [string stringByReplacingOccurrencesOfString:@"\\"
                                               withString:@"\\\\"];
    string = [string stringByReplacingOccurrencesOfString:@"\""
                                               withString:@"\\\""];
    [mutableString appendString:@"\""];
    [mutableString appendString:string];
    [mutableString appendString:@"\""];
}

+ (NSUInteger)estimateLeafNodeSize:(FLeafNode *)leafNode {
    NSString *jsType = [FUtilities getJavascriptType:leafNode.val];
    // These values are somewhat arbitrary, but we don't need an exact value so
    // prefer performance over exact value
    NSUInteger valueSize;
    if (jsType == kJavaScriptNumber) {
        valueSize = 8; // estimate each float with 8 bytes
    } else if (jsType == kJavaScriptBoolean) {
        valueSize = 4; // true or false need roughly 4 bytes
    } else if (jsType == kJavaScriptString) {
        valueSize = 2 + [leafNode.val length]; // add 2 for quotes
    } else {
        [NSException raise:NSInvalidArgumentException
                    format:@"Unknown leaf type: %@", leafNode];
        return 0;
    }

    if (leafNode.getPriority.isEmpty) {
        return valueSize;
    } else {
        // Account for extra overhead due to the extra JSON object and the
        // ".value" and ".priority" keys, colons, comma
        NSUInteger leafPriorityOverhead = 2 + 8 + 11 + 2 + 1;
        return leafPriorityOverhead + valueSize +
               [FSnapshotUtilities estimateLeafNodeSize:leafNode.getPriority];
    }
}

+ (NSUInteger)estimateSerializedNodeSize:(id<FNode>)node {
    if ([node isEmpty]) {
        return 4; // null keyword
    } else if ([node isLeafNode]) {
        return [FSnapshotUtilities estimateLeafNodeSize:node];
    } else {
        NSAssert([node isKindOfClass:[FChildrenNode class]],
                 @"Unexpected node type: %@", [node class]);
        __block NSUInteger sum = 1; // opening brackets
        [((FChildrenNode *)node) enumerateChildrenAndPriorityUsingBlock:^(
                                     NSString *key, id<FNode> child,
                                     BOOL *stop) {
          sum += key.length;
          sum +=
              4; // quotes around key and colon and (comma or closing bracket)
          sum += [FSnapshotUtilities estimateSerializedNodeSize:child];
        }];
        return sum;
    }
}

@end

File: FirebaseDatabaseSwiftCore/Sources/FSnapshotUtilities.swift
//
//  File.swift
//  
//
//  Created by Morten Bek Ditlevsen on 11/10/2021.
//

import Collections
import Foundation

let kFirebaseMaxObjectDepth = 1000
let kPayloadMetadataPrefix = "."

@objc public class FSnapshotUtilities: NSObject {
    @objc public static func nodeFrom(_ val: Any?) -> FNode {
        FSnapshotUtilitiesSwift.nodeFrom(val, priority: nil)
    }

    @objc public static func nodeFrom(_ val: Any?, withValidationFrom fn: String) -> FNode {
        FSnapshotUtilitiesSwift.nodeFrom(val, withValidationFrom: fn)
    }
    @objc public static func nodeFrom(_ val: Any?, priority: Any?) -> FNode {
        FSnapshotUtilitiesSwift.nodeFrom(val, priority: priority)
    }
    @objc public static func nodeFrom(_ val: Any?, priority: Any?, withValidationFrom fn: String) -> FNode {
        FSnapshotUtilitiesSwift.nodeFrom(val, priority: priority, withValidationFrom: fn)
    }

    @objc public static func appendHashV2Representation(for string: String, to mutableString: NSMutableString) {
        var mutable: String = String(mutableString)
        FSnapshotUtilitiesSwift.appendHashV2Representation(for: string, to: &mutable)
        mutableString.setString(mutable)
    }

    @objc public static func appendHashRepresentationV2ForLeafNode(_ node: FNode, to mutableString: NSMutableString) {
        var mutable: String = String(mutableString)
        FSnapshotUtilitiesSwift.appendHashRepresentation(for: node, to: &mutable, hashVersion: .v2)
        mutableString.setString(mutable)
    }

    @objc public static func compoundWriteFromDictionary(_ values: NSDictionary, withValidationFrom fn: String) -> FCompoundWrite {
        var compoundWrite = FCompoundWrite.emptyWrite
        var updatePaths: [FPath] = []
        for keyId in values.allKeys {
            let value = values[keyId]
            let key = FValidationSwift.validateFrom(fn, validUpdateDictionaryKey: keyId, withValue: value as Any)
            let path = FPath(with: key)
            let node = FSnapshotUtilitiesSwift.nodeFrom(value, withValidationFrom: fn)
            updatePaths.append(path)
            compoundWrite = compoundWrite.addWrite(node, atPath: path)
        }
        // Check that the update paths are not descendants of each other.
        updatePaths.sort { a, b in
            a.compare(b) == .orderedAscending
        }
        var prevPath: FPath? = nil
        for path in updatePaths {
            if let prev = prevPath, prev.contains(path) {
                fatalError("(\(fn)) Invalid path in object. Path (\(prev)) is an ancestor of (\(path)).")
            }
            prevPath = path
        }
        return compoundWrite
    }

    // Move to enum and remove this once swift conversion of usage points is done
    #warning("TODO - MOVE")
    @objc public static func estimateSerializedNodeSize(_ node: FNode) -> Int {
        if node.isEmpty {
            return 4 // null keyword
        } else if node.isLeafNode() {
            return estimateLeafNodeSize(node)
        } else if let childrenNode = node as? FChildrenNode {
            var sum = 1 // opening brackets
            for (key, child) in childrenNode.children {
                sum += key.count
                sum += 4 // quotes around key and colon and (comma or closing bracket)
                sum += estimateSerializedNodeSize(child)
            }
            return sum

        } else {
            assert(false, "Unexpected node type: \(type(of: node))")
            return 0
        }
    }
    // Move to enum and remove this once swift conversion of usage points is done
    #warning("TODO - MOVE")
    static func estimateLeafNodeSize(_ node: FNode) -> Int {
        // These values are somewhat arbitrary, but we don't need an exact value so
        // prefer performance over exact value
        let valueSize: Int
        switch FUtilitiesSwift.getJavascriptType(node.val()) {
        case .number:
            valueSize = 8 // estimate each float with 8 bytes
        case .boolean:
            valueSize = 4 // true or false need roughly 4 bytes
        case .string:
            // If we are measuring bytes here then we should use the utf8 view here, right?
            valueSize = 2 + ((node.val() as? String)?.utf8.count ?? 0) // add 2 for quotes
        default:
            fatalError("Unknown leaf type: \(node)")
        }
        if node.getPriority().isEmpty {
            return valueSize
        } else {
            // Account for extra overhead due to the extra JSON object and the
            // ".value" and ".priority" keys, colons, comma
            let leafPriorityOverhead = 2 + 8 + 11 + 2 + 1;
            return leafPriorityOverhead + valueSize +
            estimateLeafNodeSize(node.getPriority())
        }
    }
}

public enum FSnapshotUtilitiesSwift {
    enum FDataHashVersion {
        case v1
        case v2
    }

    public static func nodeFrom(_ val: Any?) -> FNode {
        nodeFrom(val, priority: nil)
    }

    public static func nodeFrom(_ val: Any?, priority: Any?) -> FNode {
        nodeFrom(val, priority: priority, withValidationFrom: "nodeFrom:priority:")
    }

    public static func nodeFrom(_ val: Any?, withValidationFrom fn: String) -> FNode {
        var path: [String] = []
        return nodeFrom(val, priority: nil, withValidationFrom: fn, atDepth: 0, path: &path)
    }

    public static func nodeFrom(_ val: Any?, priority: Any?, withValidationFrom fn: String) -> FNode {
        var path: [String] = []
        return nodeFrom(val, priority: priority, withValidationFrom: fn, atDepth: 0, path: &path)
    }

    public static func nodeFrom(_ val: Any?, priority: Any?, withValidationFrom fn: String, atDepth depth: Int, path: inout [String]) -> FNode {
        internalNodeFrom(val, priority: priority, withValidationFrom: fn, atDepth: depth, path: &path)
    }

     public static func internalNodeFrom(_ val: Any?, priority: Any?, withValidationFrom fn: String, atDepth depth: Int, path: inout [String]) -> FNode {
         guard depth <= kFirebaseMaxObjectDepth else {
             let pathString = path[0..<100].joined(separator: ".")
             fatalError("(\(fn)) Max object depth exceeded: \(pathString)...")
         }
         if val == nil || (val as? NSNull) === NSNull() {
             return FEmptyNode.emptyNode
         }
         var value = val
         FValidationSwift.validateFrom(fn, isValidPriorityValue: priority as Any, withPath: path)
         var priority = FSnapshotUtilitiesSwift.nodeFrom(priority)
         var isLeafNode = false
         if let dict = val as? NSDictionary {
             if let rawPriority = dict[kPayloadPriority] {
                 FValidationSwift.validateFrom(fn, isValidPriorityValue: rawPriority, withPath: path)
                 priority = nodeFrom(rawPriority)
             }
             if let payload = dict[kPayloadValue] {
                 value = payload
                 if FValidationSwift.validateFrom(fn, isValidLeafValue: value, withPath: path) {
                     isLeafNode = true
                 } else {
                     fatalError("(\(fn)) Invalid data type used with .value. Can only use NSString and NSNumber or be null. Found \(type(of: value)) instead.")
                 }
             }
         }
         if !isLeafNode && FValidationSwift.validateFrom(fn, isValidLeafValue: value, withPath: path) {
             isLeafNode = true
         }

         if isLeafNode {
             return FLeafNode(value: value as Any, withPriority: priority)
         }

         // Unlike with JS, we have to handle the dictionary and array cases
         // separately.

         if let dval = value as? NSDictionary {
             var children: [String: FNode] = .init(minimumCapacity: dval.count)

             // Avoid creating a million newPaths by appending to old one
             for keyId in dval.allKeys {
                 let key = FValidationSwift.validateFrom(fn, validDictionaryKey: keyId, withPath: path)
                 if !key.hasPrefix(kPayloadMetadataPrefix) {
                     path.append(key)
                     let childNode = nodeFrom(dval[key], priority: nil, withValidationFrom: fn, atDepth: depth + 1, path: &path)
                     path.removeLast()
                     if !childNode.isEmpty {
                         children[key] = childNode
                     }
                 }
             }
             if children.isEmpty {
                 return FEmptyNode.emptyNode
             } else {
                 var dict = OrderedDictionary(uncheckedUniqueKeys: children.keys, values: children.values)
                 dict.sort { a, b in
                     FUtilitiesSwift.compareKey(a.key, b.key) == .orderedAscending
                 }

                 return FChildrenNode(priority: priority, children: dict)
             }
         } else if let aval = value as? NSArray {
             var children: [String: FNode] = .init(minimumCapacity: aval.count)

             for i in 0..<aval.count {
                 let key = "\(i)"
                 path.append(key)
                 let childNode = nodeFrom(aval[i], priority: nil, withValidationFrom: fn, atDepth: depth + 1, path: &path)
                 path.removeLast()

                 if !childNode.isEmpty {
                     children[key] = childNode
                 }
             }

             if children.isEmpty {
                 return FEmptyNode.emptyNode
             } else {
                 var dict = OrderedDictionary(uncheckedUniqueKeys: children.keys, values: children.values)
                 dict.sort { a, b in
                     FUtilitiesSwift.compareKey(a.key, b.key) == .orderedAscending
                 }

                 return FChildrenNode(priority: priority, children: dict)
             }
         } else {
             let pathString = path.prefix(50).joined(separator: ".")
             fatalError("(\(fn)) Cannot store object of type \(type(of: value)) at \(pathString). Can only store objects of type NSNumber, NSString, NSDictionary, and NSArray.")
         }
     }

    static func validatePriorityNode(_ priorityNode: FNode) {
        if priorityNode.isLeafNode() {
            let val = priorityNode.val()
            if let valDict = val as? NSDictionary {
                assert(valDict[kServerValueSubKey] != nil, "Priority can't be object unless it's a deferred value")
            } else {
                let jsType = FUtilitiesSwift.getJavascriptType(val)
                assert(jsType == .string || jsType == .number, "Priority of unexpected type.")
            }
        } else {
            assert (priorityNode === FMaxNode.maxNode || priorityNode.isEmpty, "Priority of unexpected type.")
        }
        // Don't call getPriority() on MAX_NODE to avoid hitting assertion.
        assert (priorityNode === FMaxNode.maxNode || priorityNode.getPriority().isEmpty, "Priority nodes can't have a priority of their own.")
    }

    static func appendHashRepresentation(for leafNode: FNode, to output: inout String, hashVersion: FDataHashVersion) {
        if !leafNode.getPriority().isEmpty {
            output += "priority:"
            appendHashRepresentation(for: leafNode.getPriority(),
                                        to: &output,
                                        hashVersion: hashVersion)
            output += ":"
        }
        let jsType = FUtilitiesSwift.getJavascriptType(leafNode.val())
        output += jsType.rawValue + ":"
        switch jsType {
        case .object:
            fatalError("Unknown value for hashing: \(leafNode)")

        case .boolean:
            let numberVal = (leafNode.val() as? NSNumber) ?? NSNumber(booleanLiteral: false)
            output += numberVal.boolValue ? "true" : "false"
        case .number:
            let numberVal = (leafNode.val() as? NSNumber) ?? NSNumber(integerLiteral: 0)

            output += FUtilitiesSwift.ieee754String(for: numberVal)
        case .string:
            let stringVal = (leafNode.val() as? String) ?? ""
            switch hashVersion {
            case .v1:
                output += stringVal
            case .v2:
                appendHashV2Representation(for: stringVal, to: &output)
            }
        case .null:
            ()
        }
    }

    static func appendHashV2Representation(for string: String, to output: inout String) {
        output += "\""
        output += string
            .replacingOccurrences(of: "\\", with: "\\\\")
            .replacingOccurrences(of: "\"", with: "\\\"")
        output += "\""
    }

}
